llama-index-packs/llama-index-packs-longrag/llama_index/packs/longrag/base.py

```diff
@@ -0,0 +1,312 @@
+        top_parents_set: t.Set[str] = set()
+        top_parents: t.List[NodeWithScore] = []
+        for id_, similarity in zip(query_res.ids, query_res.similarities):
+            cur_node = self._small_toks_dict[id_]
+            parent_id = cur_node.metadata["parent_node"]
+            if parent_id not in top_parents_set:
+                top_parents_set.add(parent_id)
+
+                parent_node = self._grouped_nodes_dict[parent_id]
+                node_with_score = NodeWithScore(node=parent_node, score=similarity)
+                top_parents.append(node_with_score)
+
+                if len(top_parents_set) >= self._similarity_top_k:
+                    break
+
+        assert len(top_parents) == min(self._similarity_top_k, len(self._grouped_nodes))
+
+        return top_parents
+
+
+class LongRAGPack(BaseLlamaPack):
+    """Implements Long RAG.
+
+    This implementation is based on the following paper: https://arxiv.org/pdf/2406.15319
+    """
+
+    def __init__(
+        self,
+        data_dir: str,
+        llm: t.Optional[LLM] = None,
+        embed_model: t.Optional[BaseEmbedding] = None,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+        similarity_top_k: int = DEFAULT_TOP_K,
+        small_chunk_size: int = DEFAULT_SMALL_CHUNK_SIZE,
+    ):
+        """Constructor.
+
+        Args:
+            data_dir (str): Data directory
+            llm (t.Optional[LLM]): LLM
+            embed_model(Optional[BaseEmbedding]): embed model
+            chunk_size (int, optional): Splits each doc to chunk_size to demonstrate grouping. Set to -1 to disable splitting then grouping. Defaults to DEFAULT_CHUNK_SIZE.
+            similarity_top_k (int, optional): Top k. Defaults to DEFAULT_TOP_K.
+            small_chunk_size (int, optional): Small chunk size to split large documents into smaller embeddings of small_chunk_size. Defaults to DEFAULT_SMALL_CHUNK_SIZE.
+        """
+        # initialize vars
+        self._data_dir = data_dir
+        self._embed_model = embed_model or embed_model_from_settings_or_context(
+            Settings, None
+        )
+        self._llm = llm or llm_from_settings_or_context(Settings, None)
+        self._chunk_size = chunk_size
+        self._similarity_top_k = similarity_top_k
+        self._small_chunk_size = small_chunk_size
+
+        # read docs
+        docs = SimpleDirectoryReader(self._data_dir).load_data()  # read documents
+        if self._chunk_size != -1:
+            nodes = split_doc(
+                self._chunk_size, docs
+            )  # split documents into chunks of chunk_size
+            grouped_nodes = get_grouped_docs(
+                nodes
+            )  # get list of nodes after grouping (groups are combined into one node), these are long retrieval units
+        else:
+            grouped_nodes = docs
+
+        # make retriever and query engine
+        self._retriever = LongRAGRetriever(
```

Comment:
it's not a bad idea to define a custom retriever, but would be ideal is the "small" chunks were stored within a `VectorStoreIndex` because then it gives you all the integrations with different vector stores.

you could store small chunks in a vectorstoreindex, store the parent big chunks somewhere else, and then define a custom retriever to then map these small chunks into the big chunks before you return the `NodeWithScore` list